src/NHibernate/AdoNet/GenericBatchingBatcher.cs

```diff
@@ -0,0 +1,235 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.Common;
+using System.Text;
+using NHibernate.AdoNet.Util;
+using NHibernate.Exceptions;
+using NHibernate.SqlCommand;
+
+namespace NHibernate.AdoNet
+{
+	/// <summary>
+	/// A generic batcher that will batch UPDATE/INSERT/DELETE commands by concatenating them with a semicolon.
+	/// Use this batcher only if there are no dedicated batchers in the given environment. Unfortunately some
+	/// database clients do not support concatenating commands with a semicolon. Here are the known clients
+	/// that do not work with this batcher:
+	/// - FirebirdSql.Data.FirebirdClient
+	/// - Oracle.ManagedDataAccess
+	/// - System.Data.SqlServerCe
+	/// </summary>
+	public partial class GenericBatchingBatcher : AbstractBatcher
+	{
+		private readonly int? _maxNumberOfParameters;
+		private readonly BatchingCommandSet _currentBatch;
+		private int _totalExpectedRowsAffected;
+		private StringBuilder _currentBatchCommandsLog;
+
+		public GenericBatchingBatcher(ConnectionManager connectionManager, IInterceptor interceptor, char statementTerminator)
+			: base(connectionManager, interceptor)
+		{
+			BatchSize = Factory.Settings.AdoBatchSize;
+			StatementTerminator = statementTerminator;
+			_currentBatch = new BatchingCommandSet(this);
+			_maxNumberOfParameters = Factory.Dialect.MaxNumberOfParameters;
+
+			// We always create this, because we need to deal with a scenario in which
+			// the user change the logging configuration at runtime. Trying to put this
+			// behind an if(log.IsDebugEnabled) will cause a null reference exception 
+			// at that point.
+			_currentBatchCommandsLog = new StringBuilder().AppendLine("Batch commands:");
+		}
+
+		public char StatementTerminator { get; }
+
+		public sealed override int BatchSize { get; set; }
+
+		protected override int CountOfStatementsInCurrentBatch => _currentBatch.CountOfCommands;
+
+		public override void AddToBatch(IExpectation expectation)
+		{
+			var batchUpdate = CurrentCommand;
+			if (_maxNumberOfParameters.HasValue && 
+				_currentBatch.CountOfParameters + CurrentCommand.Parameters.Count > _maxNumberOfParameters)
+			{
+				ExecuteBatchWithTiming(batchUpdate);
+			}
+			_totalExpectedRowsAffected += expectation.ExpectedRowCount;
+			Driver.AdjustCommand(batchUpdate);
+			string lineWithParameters = null;
```

Comment:
Can logging (L59-74) please be moved to a separate method?

Reply:
Done.